Settings refactor V2

fwsmit opened the pull request:
Our current way of dealing with settings is very messy. It also involves copy and pasting a considerate amount of code for adding a setting. This PR strives to make it so that you only need to add a few lines to an array, `allowed_settings`, to add a setting or rule.
This allows us to do much better checking of the settings both in the way settings are defined and the user's input.
If you make a mistake in defining a setting, this will most likely be caught by the tests, or you can easily add a test to account for this mistake. We can also easily extend the set of allowed rules without making the code more messy.
If a users misspells a setting in their dunstrc, instead of ignoring that setting dunst will warn the user that setting doesn't exist. This could also be extended to suggesting a similar setting from the list of settings.

Changes in this PR
- refactor the settings code
- refactor the rules code
- refactor the command line parsing
- remove all the now unused functions: `string_parse_*`, `ini_get_*`, `option_get_*` (`cmd_line_get_*` stays for now)
- fix segfault in `string_to_color`
- multiple sections with the same name are merged
- rules don't respect global settings anymore. This makes the settings more consistent. See also:https://github.com/dunst-project/dunst/issues/644#issuecomment-517405127
- rules can be placed in the global
- Fixes #644, by allowing rules in the global section.
- warnings are emitted when an unknown option is passed

Deprecated by this PR
- deprecate some settings which were hard to implement or not needed: all the settings in the section `[shortcuts]`, `verbosity`, `icon_folders`, `startup_notification`, something I forgot?. (`startup_notification` and `verbosity` are still command line options)
- deprecate setting settings via command line arguments (#328)
- deprecate STATIC_CONFIG (#414)

TODO:
- [x] Add warning for deprecated settings, instead of the default 'does not exists'
- [x] Set the same defaults as before
- [x] Fix warnings about unused variables in settings_data.h
- [x] Add some more tests, especially some tests with a test dunstrc

Possible enhancements:
- Warn when the same setting is set multiple times in the same section

Implements: #412 #414 
Fixes: #138

bebehei reviewed (commented):
Nice PR. 

You might want to look into #475

I'm not sure about the current state of option parsing in dunst, but I'd prefer to have uniformed string parser functions. This can be used then in further work to restructure the option parser.

fwSmit commented on src/option_parser.c:
```diff
@@ -354,6 +371,150 @@ int ini_get_bool(const char *section, const char *key, int def)
+                .parser = string_parse_alignment_2,
+        },
+        {
+                // Last entry is empty so the code knows where the array ends
```
Ah tanks for that. I was looking for that

fwSmit commented on src/settings.h:
```diff
@@ -101,6 +103,30 @@ struct settings {
+      char *name; //config key
+      char *section; //config section
+      // TODO implement cmdopt
+      char *cmdopt; // cmdline option
```
I agree. We should still provide some command line options though. Or do you think none of the settings need command line options? What does @tsipinakis think about it?
The command line options definitely need a refactor though, because now, there's no way to detect if an unknown option is passed.

fwSmit commented on src/option_parser.c:
```diff
@@ -54,6 +54,24 @@ bool string_parse_alignment(const char *s, enum alignment *ret)
         return false;
 }
 
+int string_parse_alignment_2(const char *s, void *ret)
```
I think we can get rid of most of the string parse functions and just replace them with one. Just pass an array with the possible strings linked to the enum value it belongs to.

fwSmit commented:
> I'm not sure about the current state of option parsing in dunst, but I'd prefer to have uniformed string parser functions. This can be used then in further work to restructure the option parser.

Do you mean the same thing I mentioned here https://github.com/dunst-project/dunst/pull/803#discussion_r554405827? Then I'm totally down.

> Nice PR.
> 
> You might want to look into #475

That's a good PR too. I wonder why it never got merged. There are quite a few small things in there that make the settings code better. But before we continue this implementation in C I think it's a good time to decide 2 things:
- Do we want to continue this in C or port the entire settings code over to Rust.
- I don't see any issues about this in this repo so it might be a bit short notice, but it's said a few times that we've outgrown the ini configuration format. Do we want to think about moving to a new configuration format entirely?

fwSmit commented:
This PR is almost ready, but I don't know what makes the last tests fail. A second pair of eyes would help a lot.
Restoring settings.c and config.h to the state from before the PR makes more dbus tests fail. 
@bebehei could you take a look, since you made the dbus tests.

fwsmit committed:
Add generic option parser for settings refactor V2

This is implemented by giving each setting a type. Based on the type,
the string will be parsed differently.
A generic enum parser is used as well. Instead of having a function for
every enum, only one function is needed. This will take some data with
possible enum values.

All data needed to parse the settings is defined in settings_data.h.
This is useful for keeping all data in one place and it's easier to
check if there are mistakes.

fwsmit committed:
Add TYPE_LIST and TYPE_SEP_COLOR parsers

fwsmit committed:
Add tests for settings_data.h

fwsmit committed:
Add data for all settings and enums

This data is extracted from src/settings.c with help from awk and vim
macros. The enum data is extracted from all the string_parse_* functions
in src/option_parser.c.
For future reference, here's the awk script I used

        #!/usr/bin/awk -f

        BEGIN {
                ACTIVE = 0
                IN_SETTING = 0
                SETTING_TYPE = "unknown"
                SETTING_LOCATION = "unknown"
                SETTING_NAME = "unknown"
                SETTING_SECTION = "unknown"
                SETTING_DEFAULT = "unknown"
                SETTING_DESCRIPTION = "unknown"
                errors = ""
                unextracted = ""
                n_lines = 0
                line_number = 64
        }

        # Start of extraction
        /^void load_settings/ { ACTIVE = 1 }

        # end of extraction
        /push hardcoded default rules into rules list/ { ACTIVE = 0 }

        /settings\..* = option_get/ {
                if (ACTIVE) {
                        if (IN_SETTING == 0){
                                IN_SETTING = 1
                                SETTING_LOCATION = $1
                                if ($3 == "option_get_bool("){
                                        SETTING_TYPE = "TYPE_BOOLEAN"
                                } else if ($3 == "option_get_string("){
                                        SETTING_TYPE = "TYPE_STRING"
                                } else if ($3 == "option_get_time("){
                                        SETTING_TYPE = "TYPE_TIME"
                                } else if ($3 == "option_get_int("){
                                        SETTING_TYPE = "TYPE_INT"
                                } else if ($3 == "option_get_path("){
                                        SETTING_TYPE = "TYPE_PATH"
                                } else {
                                        SETTING_TYPE = "TYPE_NOT_FOUND"
                                        errors = errors "TYPE_NOT_FOUND " $3 " line " line_number "\n"
                                }
                                n_lines++
                        }
                }
        }

        {
                if (IN_SETTING){
                        if (n_lines == 2) {
                                SETTING_SECTION = substr($1, 0, length($1)-1)
                        }
                        if (n_lines == 3) {
                                SETTING_NAME = substr($1, 0, length($1)-1)
                                n = split($0, array, ",")
                                SETTING_DEFAULT = array[3]
                                sub(/^ +/, "", SETTING_DEFAULT)
                        }
                        if (n_lines == 4) {
                                SETTING_DESCRIPTION = $0
                                sub(/^ +/, "", SETTING_DESCRIPTION)
                        }
                     # print
                     n_lines ++
                }
                line_number ++
                if (ACTIVE && !IN_SETTING) {
                        unextracted = unextracted $0 "\n"
                }
        }

        # end of setting
        /^        );$/ {
                IN_SETTING = 0
                if (n_lines != 6) {
                        errors = errors "ERROR, not the expected number of lines (" n_lines ") line " line_number "-" line_number+n_lines "\n"
                } else {
                        print "{"
                        print "        .name = " SETTING_NAME ","
                        print "        .section = " SETTING_SECTION ","
                        print "        .description = " SETTING_DESCRIPTION ","
                        print "        .type = " SETTING_TYPE ","
                        print "        .value = " SETTING_LOCATION ","
                        print "        .default_value = " SETTING_DEFAULT ","
                        print "        .parser = NULL,"
                        print "        .parser_data = NULL,"
                        print "},"
                }

                SETTING_TYPE = "unknown"
                SETTING_LOCATION = "unknown"
                SETTING_NAME = "unknown"
                SETTING_SECTION = "unknown"
                SETTING_DEFAULT = "unknown"
                SETTING_DESCRIPTION = "unknown"
                n_lines = 0
        }

        END {
                print "These lines have not been extracted"
                print unextracted
                print "These errors have to be manually corrected: "
                print errors
        }

fwsmit committed:
Deprecate STATIC_CONFIG

As said in #414, STATIC_CONFIG is a very old feature and seems to be
used by almost nobody.

> As usual if someone wants to use dunst without a configuration option,
> using dunst -config /dev/null should be sufficient.

fwsmit committed:
Start using the new settings parser

This commit also removes config.h, since it's not used anymore. All
defaults can be set in settings_data.h.

fwsmit committed:
Replace free_string_array with g_strfreev

Since we already depend on glib, we might as well use it.
The two functions are almost exactly the same. I looked at the source
code of glib.

fwsmit committed:
Fix segfault when passing NULL to string_to_color

fwsmit committed:
Remove ununsed string_parse_* functions and rename the newly created ones.

fwsmit committed:
Remove unused option_get_* functions

fwsmit committed:
Add a list of special sections

This will prevent warnings and will allow to implement rules with the
new parser later on.

fwsmit committed:
Add a get_rule function

fwsmit committed:
Change the way the default rule is set

fwsmit committed:
rules: implement rules with the new settings system

fwsmit committed:
parser: remove the ini_get_* functions

They're unused and superseeded by settings V2

fwSmit commented:
This PR is finally in a reviewable state. I do not expect everything to work perfectly yet. There's probably some things that have changed from the old version. I will go through and test a bit more. The code is mostly ready, so you can go ahead and review it.

fwsmit committed:
docs: update the documentation regarding the rule changes.

fwsmit committed:
settings: rulify the settings

This commit tries to make the setttings more consistent by implementing
the settings as rules everywhere where it's possible. This means the
rule code from settings.c can be removed and everything is defined in
settings_data.h. Rules are now also allowed in the global and urgency
sections. This means #644 is now fixed.

The special sections are implemented as rules by giving them implied
filters. The global sections has no filters and the urgency sections
only have filters of their respective urgency. To avoid confusion, no
other filters are allowed in these sections.

Fixes #644

fwsmit committed:
settings: remove startup_notification as a setting

It can still be used from the command line

fwsmit committed:
dbus: init the notification before modifying some attributes

Since there are some rules set by default now, I discovered that the
dbus code doesn't handle them correctly. It set the colors and some
other things before initializing the notification. These changes were
overwritten by the rules, which probably isn't desired.

On another note, who actually uses this feature?

fwsmit committed:
tests: fix test/queues

This one also didn't account for rules being there

fwsmit committed:
test: add tests for set_from_string and fix it's problems

Most problems appear when setting the same value multiple times. Fixed
the frees and made checks more thorough.

fwsmit committed:
build: fix all warnings on gcc compilers

The warning avoided here is -Wunused-variable. It's even triggered when
the variable is used in on of the compilation units. By making the
variables static const, this only happens if they're unused in the main
compilation unit. (see gcc(1))

fwsmit committed:
build: avoid clang warning

Avoid -Wtautological-constant-out-of-range-compare warning from clang.
This means you are relying on implementation-defined behaviour of the
enum type. It's better to include the -1 in the enum itself to make it
more clear and avoid the enum being an incompatible type. See
https://stackoverflow.com/questions/42344476/clang-tautological-constant-out-of-range-compare-warning#42344912

fwsmit committed:
test: add invalid_int test and test for variables not changing from invalid values

fwsmit committed:
Remove remaining string_parse_* functions

And one more unused function

fwsmit committed:
settings: set the same defaults as before the refactor

The only exception is not setting idle_threshold, since I don't think
many people want to use it and the docs say the default is 0.

fwsmit committed:
wayland: remove warning when settings.idle_threshold is 0

fwsmit committed:
settings: use DIE instead of LOG_E on invalid setting

LOG_E gives the impression that the program crashed. This should not
happen for invalid user input, only for unexped situations not caused by
user input.

fwsmit committed:
test: add tests with some dunstrcs

Also check if the defaults match between settings_data.h and the
default dunstrc. They should be exactly the same.

fwsmit committed:
settings: Generalize list code for adding more types of lists

fwsmit committed:
cmd: Add the -print command line argument again

Also add format to the printed attributes

fwsmit committed:
option_parser: refactor variables in set_from_string

fwsmit committed:
settings_data: replace TYPE_ENUM with TYPE_CUSTOM

fwsmit committed:
settings: replace TYPE_SEP_COLOR with TYPE_CUSTOM

fwsmit committed:
test/setting: add a comment explaining the check

fwsmit committed:
settings_data: replace TYPE_BOOLEAN with TYPE_CUSTOM

fwSmit commented:
@bebehei The tests are now also compiled with -Werror and that's making it harder to write tests (I'm using a language extension supported by clang and gcc, but clang gives a warning about it). I think it's best to compile only dunst itself with -Werror. I tried implementing that, but now the code coverage is giving errors. If you have time, could you take a look at that?

@tsipinakis Other than that, the PR is ready. You can take a look at the changes  and I especially recommend looking at the first comment in this PR for a list of changes. Let me know if you don't agree with any of the choices I made

fwsmit committed:
test/settings: remove the use of typeof to avoid a clang warning

fwsmit committed:
test/option_parser: avoid -Wtautological-constant-out-of-range-compare warning from clang

tsipinakis commented:
For the record, I've already started a review on this. About half-way through but life keeps getting in the way (and this is a huge PR).
No comments so far, amazing work!

tsipinakis commented:
It has taken.. official 6 months. Wow sorry for this :see_no_evil:. This has been a gigantic leap forward in the code quality and readability. Thanks a lot @fwSmit.. Now to review the others :)

fwSmit commented:
> It has taken.. official 6 months. Wow sorry for this see_no_evil. This has been a gigantic leap forward in the code quality and readability. Thanks a lot @fwSmit.. Now to review the others :)

Thanks for reviewing. Those 6 months weren't all you, I was still developing the branch as well ;)
